### DISCUSSION TITLE
Validation error in LDPC5GDecoder

### DISCUSSION THREAD

**Discussion by invicta198** [NVlabs/sionna#531] [Q&A]
I am reaching out to request assistance regarding an issue I am encountering with [LDPC5GDecoder](https://github.com/NVlabs/sionna/blob/0a4a22ec484d80ed038bf9b30240dfa4a3f90216/sionna/fec/ldpc/decoding.py#L992).

I am performing some tests over a 5G Accelerator, ACC200, using [DPDK-BBdev](https://github.com/DPDK/dpdk/tree/v22.11/app/test-bbdev). For the same, I am creating custom test-vector files to generate custom load over the ACC200 device. I am creating the test-vector files with help of sionna. 

Following are the steps to create **input0** and **output0** used in BBdev vector file. I have pulled out this code from the following reference: https://github.com/NVlabs/sionna/discussions/54.

```
k, n, q_m = 32, 64, 2
encoder = sn.fec.ldpc.LDPC5GEncoder(k, n, num_bits_per_symbol=q_m)
c = encoder(u)

e = c.numpy()

for i in range(e.shape[0]):
    for j in flipped_positions[i]:
        e[i][j] = 1.0 if e[i][j] == 0.0 else 0.0
e = tf.convert_to_tensor(e, np.float32)
decoder = sn.fec.ldpc.LDPC5GDecoder(encoder, hard_out=True, return_infobits=True, num_iter=20, cn_type="minsum")
w = decoder(2*e - 1)
```

When I use w as **input0** and u as **output0** in BBdev vector file I face following validation error:

> Enqueue 32 reqs (phys 0x17ff80000) to reg 0x1180000000
> Debug : MMIO Enqueue
> TestCase check_dec_status_and_ordering() line 1936 failed: op_status (8) != expected_status (0)
> TestCase validate_ldpc_dec_op() line 2328 failed (err -1): Checking status and ordering for decoder failed
> TestCase latency_test_ldpc_dec() line 4645 failed (err -1): Validation failed!
> 727:get_stats_from_queues() Got stats on 0
> 762:rte_bbdev_stats_get() Retrieved stats of device 0
> 602:rte_bbdev_stop() Stopped device 0
> TestCase [ 0] : validation_tc failed
> 647:rte_bbdev_close() Closed device 0
> 

Please find below the BBdev vector file created for the same

```
op_type =
RTE_BBDEV_OP_LDPC_DEC

input0=
0xff01ff01,0x0101ff01,0x0101ff01,0x01010101,0x0101ff01,0xff010101,0x01010101,0x01010101,0xff01ff01,0x0101ff01,0xff010101,0xffffff01,0xff0101ff,0x0101ff01,0xffff0101,0xff0101ff

output0=
0xffffffff

basegraph=
2

z_c=
8

q_m=
2

n_filler=
28

n_cb=
300

e=
64

rv_index =
0

code_block_mode =
1

iter_max =
20

expected_iter_count =
20

op_flags =
RTE_BBDEV_LDPC_ITERATION_STOP_ENABLE

expected_status =
OK
```

I would like to understand what incorrect configuration am I using?

Any insights or guidance to help resolve this issue would be highly appreciated. If you need any more details, please feel free to ask.

Thank you for your time and assistance!

**Comment by SebastianCa** [comment]
Hi @invicta198,
Reproducing the results without thoroughly examining the bbdev test cases seems challenging. Could you please provide a specific NumPy version of the test case that is failing?

It seems that there is already a discrepancy in the lifting factor `Z` (or `z_c` in your example) between Sionna and your implementation. In the provided code snippet, `encoder._z=6`, while in your example it is `z_c=8`. Thus, I'm not sure if the Sionna encoder is correctly configured for the given testcase.

Is it possible that the test case evaluates performance at the transport block level? This would include an outer CRC check. For more details, please refer to the [Transport Block section in the API documentation](https://nvlabs.github.io/sionna/api/nr.html#transport-block).
